rust/agents/relayer/src/msg/processor.rs

```diff
@@ -63,92 +66,106 @@ impl MessageProcessor {
+                metrics.set(message.nonce as i64);
+            }
+
+            // If this message has already been processed, on to the next one.
+            if self
+                .db
+                .retrieve_message_processed(self.message_nonce)?
+                .is_some()
+            {
+                debug!(
+                    msg_nonce=?self.message_nonce,
+                    "Message already marked as processed in DB");
+                self.message_nonce += 1;
+                return Ok(None);
+            }
+
+            return Ok(Some(message));
+        } else {
+            debug!(
+                msg_nonce=?self.message_nonce,
+                "No message found in DB for nonce");
+            return Ok(None);
+        };
+    }
+
     /// One round of processing, extracted from infinite work loop for
     /// testing purposes.
     async fn tick(&mut self) -> Result<()> {
         // Scan until we find next nonce without delivery confirmation.
-        if self
-            .db
-            .retrieve_message_processed(self.message_nonce)?
-            .is_some()
-        {
-            debug!(
-                nonce=?self.message_nonce,
-                "Skipping since message_nonce already in DB");
-            self.message_nonce += 1;
-            return Ok(());
-        }
-        let message = if let Some(msg) = self
-            .db
-            .message_by_nonce(self.message_nonce)?
-            .map(HyperlaneMessage::from)
-        {
-            debug!(msg=?msg, "Working on msg");
-            msg
-        } else {
-            debug!("Leaf in db without message nonce: {}", self.message_nonce);
-            // Not clear what the best thing to do here is, but there is seemingly an
-            // existing race wherein an indexer might non-atomically write leaf
-            // info to rocksdb across a few records, so we might see the leaf
-            // status above, but not the message contents here.  For now,
-            // optimistically yield and then re-enter the loop in hopes that the
-            // DB is now coherent. TODO(webbhorn): Why can't we yield here
-            // instead of sleep?
-            tokio::time::sleep(Duration::from_secs(1)).await;
-            return Ok(());
-        };
-        if let Some(metrics) = self.metrics.get(message.destination) {
-            metrics.set(self.message_nonce as i64);
-        }
-
-        // Skip if not whitelisted.
-        if !self.whitelist.msg_matches(&message, true) {
-            debug!(
-                id=?message.id(),
-                destination=message.destination,
-                nonce=message.nonce,
-                whitelist=?self.whitelist,
-                "Message not whitelisted, skipping");
-            self.message_nonce += 1;
-            return Ok(());
-        }
+        if let Some(message) = self.try_get_unprocessed_message()? {
+            debug!(msg=?message, msg_id=?message.id(), msg_nonce=message.nonce, "Working on message");
```

Comment:
Does the message debug formatter not already include the id and nonce?

Reply:
I suppose it does